prov/util/src/util_ns.c

```diff
@@ -0,0 +1,635 @@
+	return FI_SUCCESS;
+}
+
+static void util_ns_name_server_cleanup(void *args)
+{
+	void **cleanup_args = (void **)args;
+	ofi_close_socket((uintptr_t)cleanup_args[0]);
+	util_ns_map_fini((struct util_ns *)cleanup_args[1]);
+}
+
+static int util_ns_op_dispatcher(struct util_ns *ns,
+				 struct util_ns_cmd *cmd,
+				 SOCKET sock)
+{
+	int ret = FI_SUCCESS;
+	size_t io_len = 0;
+	void *io_buf = NULL, *service, *name;
+
+	switch (cmd->op) {
+	case OFI_UTIL_NS_ADD:
+	case OFI_UTIL_NS_DEL:
+		io_len = ns->name_len + ns->service_len;
+		io_buf = calloc(io_len, 1);
+		if (!io_buf) {
+			ret = -FI_ENOMEM;
+			goto fn1;
+		}
+
+		ret = util_ns_read_socket_op(sock, io_buf, io_len);
+		if (ret == io_len) {
+			service = calloc(ns->service_len, 1);
+			if (!service) {
+				ret = -FI_ENOMEM;
+				goto fn2;
+			}
+			name = calloc(ns->name_len, 1);
+			if (!name) {
+				ret = -FI_ENOMEM;
+				goto fn3;
+			}
+
+			io_len = 0;
+			memcpy(service, io_buf, ns->service_len);
+			io_len += ns->service_len;
+			memcpy(name, (void *)((char *)io_buf + io_len),
+			       ns->name_len);
+
+			if (cmd->op == OFI_UTIL_NS_ADD) {
+				ret = util_ns_map_add(
+					ns, service, name
+				);
+				/* if success, no service and name
+				 * must be freed */
+				if (!ret)
+					goto fn2;
+			} else {
+				ret = util_ns_map_del(
+					ns, service, name
+				);
+				/* since name and service must be freed,
+				 * go down */
+			}
+			goto fn4;
+		} else {
+			ret = -FI_ENODATA;
+			goto fn2;
+		}
+
+	case OFI_UTIL_NS_QUERY:
+		io_len = ns->service_len;
+		io_buf = calloc(io_len, 1);
+		if (!io_buf) {
+			ret = -FI_ENOMEM;
+			goto fn1;
+		}
+
+		ret = util_ns_read_socket_op(sock, io_buf, io_len);
+		if (ret == io_len) {
+			service = calloc(ns->service_len, 1);
+			if (!service) {
+				ret = -FI_ENOMEM;
+				goto fn2;
+			}
+			memcpy(service, io_buf, ns->service_len);
+		} else {
+			ret = -FI_ENODATA;
+			goto fn2;
+		}
+
+		cmd->op = OFI_UTIL_NS_ACK;
+		cmd->status = util_ns_map_lookup(
+			ns, service, &name
+		);
+
+		free(io_buf);
+
+		if (!cmd->status) {
+			io_buf = calloc(
+				cmd_len + ns->service_len + ns->name_len, 1
+			);
+			if (!io_buf) {
+				ret = -FI_ENOMEM;
+				free(service);
+				goto fn1;
+			}
+
+			io_len = 0;
+			memcpy(io_buf, cmd, cmd_len);
+			io_len += cmd_len;
+			memcpy((void *)((char *)io_buf + io_len), service,
+			       ns->service_len);
+			io_len += ns->service_len;
+			memcpy((void *)((char *)io_buf + io_len), name,
+			       ns->name_len);
+			io_len += ns->name_len;
+		} else {
+			io_len = cmd_len;
+			io_buf = calloc(io_len, 1);
+			if (!io_buf) {
+				ret = -FI_ENOMEM;
+				free(service);
+				goto fn1;
+			}
+
+			memcpy(io_buf, cmd, cmd_len);
+		}
+
+		ret = util_ns_write_socket_op(sock, io_buf, io_len);
+		ret = ((ret == io_len) ? FI_SUCCESS : -FI_ENODATA);
+		goto fn3;
+
+	default:
+		ret = -FI_ENODATA;
+		assert(0);
+		goto fn1;
+	}
+
+fn4:
+	free(name);
+fn3:
+	free(service);
+fn2:
+	free(io_buf);
+fn1:
+	return ret;
+}
```

Comment:
The buffer allocation and free operations in this function seem to be overly entangled.  One way to simplify that is to have the ns_map_add function allocate it's local copy of name and service and have the ns_map_lookup function copy out the result instead of returning the pointer. If doing this way, the dispatch function needs only to allocate an io_buf that is big enough to hold the largest of the commands/results and sub-part of the io_buf can then be passed directly to the ns_map_xxx calls and the retrieved result can be sent directly back to the requester.